src/leap/common/http.py

```diff
@@ -36,21 +36,141 @@
+        """
+        d = _HTTP11ClientProtocol.request(self, request)
+        if self._timeout:
+            self._last_buffer_len = 0
+            timeoutCall = reactor.callLater(
+                self._timeout, self._doTimeout, request)
+            self._finishedRequest.addBoth(self._cancelTimeout, timeoutCall)
+        return d
+
+    def _doTimeout(self, request):
+        """
+        Give up the request because of a timeout.
+
+        :param request: The object defining the parameters of the request to
+                        issue.
+        :type request: twisted.web._newclient.Request
+        """
+        buffer_len = len(self._parser._buffer)
+        if not self._parser._everReceivedData \
+                or self._last_buffer_len == buffer_len:
+            self._giveUp(
+                defer.TimeoutError(
+                    "Getting %s took longer than %s seconds."
+                    % (request.absoluteURI, self._timeout)))
+        else:
+            self._last_buffer_len = buffer_len
+
+    def _cancelTimeout(self, passthrough, timeoutCall):
+        """
+        Cancel the request timeout, when it's finished.
+
+        :param passthrough: The result of previous callback.
+        :type passthrough: any
+        :param timeoutCall:
+        :type timeoutCall:
+        """
+        if timeoutCall.active():
+            timeoutCall.cancel()
+        return passthrough
+
+
+class HTTP11ClientFactory(_HTTP11ClientFactory):
+    """
+    A timeout-able HTTP 1.1 client protocol factory.
+    """
+
+    def __init__(self, quiescentCallback, timeout):
+        """
+        :param quiescentCallback: The quiescent callback to be passed to
+                                  protocol instances, used to return them to
+                                  the connection pool.
+        :type quiescentCallback: callable(Protocol)
+        :param timeout: The timeout for requests made by protocols created by
+                        this factory.
+        :type timeout: float
+        """
+        _HTTP11ClientFactory.__init__(self, quiescentCallback)
+        self._timeout = timeout
+
+    def buildProtocol(self, _):
+        """
+        Build the HTTP 1.1 client protocol.
+        """
+        return HTTP11ClientProtocol(self._quiescentCallback, self._timeout)
+
+
+class HTTPConnectionPool(_HTTPConnectionPool):
+    """
+    A timeout-able HTTP connection pool.
+    """
+
+    _factory = HTTP11ClientFactory
+
+    def __init__(self, reactor, persistent, timeout):
+        _HTTPConnectionPool.__init__(self, reactor, persistent=persistent)
+        self._timeout = timeout
+
+    def _newConnection(self, key, endpoint):
+        def quiescentCallback(protocol):
+            self._putConnection(key, protocol)
+        factory = self._factory(quiescentCallback, timeout=self._timeout)
+        return endpoint.connect(factory)
+
+
+def createPool(
+    maxPersistentPerHost=10, persistent=True, timeout=_defaultTimeout):
```

Comment:
Could you indent the parameters to make it more visible that are not part of the code of the function.

If there is no global pool (see down), this could be a private method of the HTTPClient class.